### DISCUSSION TITLE
How to mount replicated dataset without breaking further replication?

### DISCUSSION THREAD

**Discussion by Malvineous** [zrepl/zrepl#725] [General]
I've looked through the documentation but I can't find anything that mentions how to mount or otherwise access the replicated dataset without breaking the replication process.

Once before I tried mounting the filesystem but I recall this broke the replication process, because it couldn't update the dataset once it had been modified due to the mount.

Are there instructions anywhere (ideally at https://zrepl.github.io) that explains the correct way to access a replicated dataset/snapshot in such a way that the data continues to be replicated without any problems?

**Comment by InsanePrawn** [comment]
AFAIK there's no good way to have it permanently mounted.

For singular recovery purposes one can use the readonly flag and/or roll the dataset back after unmounting.
`zfs clone` also exists.


"Why can't is just have it permanently mounted in readonly mode?" - IIRC the problem here is that mounting children inside the dataset will either create a directory (modification of the dataset!) for the mountpoint or fail to do that and thus fail mounting - neither are what you want, probably.
This will of course not happen with datasets that already have that folder, but I think that's a huge sharp edge and there's probably more around.
Feel free to investigate this, the findings might be interesting for the zrepl docs.

**Comment by rkelleyrtp** [comment]
Started working with zrepl last night and found a way to keep the snapshot permanently mounted and visible on the backup server.  This works great on Debian 12 with ZFS 2.1.11 and zrepl 0.61.  

The trick:
* Ensure the dataset on the target system is set to readonly mode (eg: `zfs set readonly=yes <dataset>`)
* Ensure the dataset on the target system has a mountpoint (eg: `zfs set mountpoint=<mountpoint> <dataset>`)

Once those options are set, simply mount the dataset on the target, and you will see the data.   Here are the commands I used when the remote dataset is named `mypool/zrepl-01/mypool/data-01`:
```
zfs set readonly=yes mypool/zrepl-01
zfs set readonly=yes mypool/zrepl-01/mypool
zfs set readonly=yes mypool/zrepl-01/mypool/data-01
zfs set mountpoint=/backup_data/data-01 mypool/zrepl-01/mypool/data-01
zfs mount mypool/zrepl-01/mypool/data-01
```
Adding the readonly=yes option to the parent ensures any other zrepl volume will get the same readonly setting.  The best part: the mount point always has the data from the most recent snapshot.  If you need data from another snapshot, simply go to /backup/data-01/.zfs/snapshot/<snapshot_dir_name>).

Finally, if you need a read-write copy of the data, simply do `zfs clone mypool/zrepl-01/mypool/data-01@<snapshot-name>` then mount the clone to a directory.

Hope this helps.

**Reply by problame** [reply]
Some remarks:

1. `readonly=yes` gets inherited, so, you can just set it on the `root_fs` dataset on the receiver.
2. In the past, mounting over other readonly=yes dataset caused problems for recv.

My approach is to use clones on the receiver, i.e., 

* `root_fs` has `readonly=yes` `canmount=off` `mountpoint=none`.
* For recovering a file for restore:
  * On the receiving side, pick the snapshot you want to restore from (zfs list -r -d 1 -t snapshot recvside_pool/my_root_fs/filesystem/on/the/receiver/do/restore/from`.
  * Let's call that snapshot `recvside_pool/my_root_fs/filesystem/on/the/receiver/do/restore/from@zrepl_2023-10-14-13:00:00Z`
  * `zfs clone -o mountpoint=/tmp/restore recvside_pool/my_root_fs/filesystem/on/the/receiver/do/restore/from@zrepl_2023-10-14-13:00:00Z recvside_pool/restore`
  * Recover the files from mountpoint `/tmp/restore`
  * zfs destroy `recvside_pool/restore`

While the clone exists, pruning of the recv side won't be able to delete the snapshot, but, replication will continue to work just fine.

Please try out this approach.

Would appreciate a contribution to the zrepl docs, e.g., in the quickstart guide section, that polishes up the scripts above, *from a ZFS beginner's perspective*.

**Reply by rkelleyrtp** [reply]
While I like your approach, the problem is:  the user must create clones for each snapshot until they find the file(s) needed for recovery.  Until the snapshots/clones are mounted, the user has no idea which snapshots/clones has the files.  I agree clones are the best way to get access to the data, but constantly mounting and unmounting clones is not practical.  Using the method I described, the user always has access to the directory structure and can easily locate the file(s) needed for recovery.

Is there another way of looking at the directory structure of a snapshot w/out mounting it?  Also, do you know what kind of problems are faced on the recv side when mounting snapshots?

**Reply by problame** [reply]
>  I agree clones are the best way to get access to the data, but constantly mounting and unmounting clones is not practical. Using the method I described, the user always has access to the directory structure and can easily locate the file(s) needed for recovery.

I see your point, but you're conflating two issues here.

My "clones" model gives you true Point-in-time recovery, but you have to do the cloning.

Your model just gives users access to the latest snapshot.
Or, do you allow access to the `.zfs/snapshot/*` directory , e.g., [like described here](https://www.truenas.com/community/threads/how-to-access-zfs-snapshots-over-smb.69449/).

> Is there another way of looking at the directory structure of a snapshot w/out mounting it? Also, do you know what kind of problems are faced on the recv side when mounting snapshots?

I guess the `.zfs/snapshot/*` directory gets you pretty close.
But yeah, one would need to follow your method with `readonly=` + `mountpoint=`.
It may very well be that it works well in newer ZFS versions.

Learn more about `.zfs/snapshot` at https://openzfs.github.io/openzfs-docs/man/master/7/zfsprops.7.html#snapdir

I'd also appreciate a quickstart guide that documents your approach :)

**Reply by rkelleyrtp** [reply]
Yeah, guess I forgot to add: the recv side datasets have 'snapdir=visible' for my approach to work.  For completeness, my approach requires the following options on the recv side datasets:
* readonly=yes
* mountpoint=<some_mount_point>
* canmount=yes
* snapdir=visible

Happy to contribute something to the quick start guide.  What is the best way to submit the notes/steps?

**Reply by problame** [reply]
Best would be to write it in RST and open a PR against the zrepl docs:

https://github.com/zrepl/zrepl/tree/master/docs

If you're more familiar with markdown, a PR that adds a markdown file somewhere in the repo will also do. I can polish it up.

Please use a tutorial-style of writing, e.g.,  https://zrepl.github.io/quickstart/backup_to_external_disk.html or https://zrepl.github.io/quickstart/fan_out_replication.html

---

If you think your configuration above should be part of quickstart guides (in practice, the configs there are what most people use), feel free to change them in your PR.

---

Thank you! 🙏

**Reply by kpfleming** [reply]
I'm experimenting with applying these settings to one of my backup drives, and I find that setting `readonly=yes` at the `root_fs` level means that mountpoints can't be created as datasets for new clients are received. I think this means it will have to be set by zrepl itself as it creates the datasets, and definitely not set on placeholder datasets. Does that sound correct?

**Reply by rkelleyrtp** [reply]
@kpfleming  Interesting, I was able to send additional datasets even when `root_fs` was set to `readonly=true` on the recv (sink) side.  For clarity, I have two test machines (`zrepl-01` and `zrepl-02`).  Each server had an initial pool called `mypool`.  

The zrepl.yml file on the sender side:
```
global:
  logging:
    - type: syslog
      format: human
      level: warn

jobs:
- name: zrepl-01__data-01
  type: push
  connect:
    type: tcp
    address: "192.168.101.126:8888"
  filesystems: {
    "mypool<": true,
  }
  snapshotting:
    type: periodic
    prefix: zrepl_
    interval: 1m
  pruning:
    keep_sender:
    - type: not_replicated
    - type: last_n
      count: 3
    keep_receiver:
    - type: last_n
      count: 3
```

The zrepl.yml file on the recv (sink) side:
```
global:
  logging:
    - type: syslog
      format: human
      level: warn

jobs:
- name: zrepl-01_backup_dataset
  type: sink
  serve:
    type: tcp
    listen: ":8888"
    clients: {
       "192.168.101.125 : "zrepl-01"
    }
  root_fs: "mypool"
```

The recv side has `readonly=on` for the `mypool` (root_fs).  I was able to create multiple child datasets on the sender side (eg `zfs create mypool/data-01`, `zfs create mypool/data-02`, etc), and all the datasets were transferred to the recv side w/out issue.

I created a bunch of datasets on the sender side, and was able to get all of them on the recv side (even child datasets).  Example:
```
NAME                                 USED  AVAIL     REFER  MOUNTPOINT
mypool                              5.32G  91.1G       96K  /mypool
mypool/zrepl-01                     5.28G  91.1G       96K  none
mypool/zrepl-01/mypool              5.28G  91.1G      112K  none
mypool/zrepl-01/mypool/data-01      1.17G  91.1G     1.17G  /data/01
mypool/zrepl-01/mypool/data-01/test    96K  91.1G       96K  /data/01/ron
mypool/zrepl-01/mypool/data-02       805M  91.1G      805M  /data/02
mypool/zrepl-01/mypool/data-03       603M  91.1G      603M  /data/03
mypool/zrepl-01/mypool/data-04       801M  91.1G      801M  /data/04
mypool/zrepl-01/mypool/data-05       986M  91.1G      986M  /data/05
mypool/zrepl-01/mypool/data-06      1004M  91.1G     1004M  /data/06
mypool/zrepl-01/mypool/data-07        96K  91.1G       96K  /data/07
mypool/zrepl-01/mypool/data-08        96K  91.1G       96K  /data/08
```

How does this compare with your config files?

**Reply by kpfleming** [reply]
Ahh, I see the difference. You are allowing the mountpoints from the sender to flow through during receive, which I don't because it would cause conflicts (all of my senders have `rpool` filesystems with mountpoints of `/`). I have zrepl configured to forcibly inherit the mountpoint property when it creates new datasets, and as a result on my system `mypool/zrepl-01` would be mounted at something like `/backups/zrepl-01` and read-only, so creation of `mypool/zrepl-01/mypool` would experience an error creating the mountpoint for it.

**Reply by rkelleyrtp** [reply]
To be clear on my test systems, the mountpoint variables are set on the receiver side (outside of zrepl) once the initial zrepl sync is done.  In fact, the mountpoints listed above are not the "real" mountpoint directory on the source system.

**Reply by kpfleming** [reply]
Thanks for the clarification. I don't do that :-)

**Reply by kpfleming** [reply]
In any case, I think this is all fine. The guide should not recommend setting `readonly=yes` at the `root_fs` level, but should instead recommend using `recv` -> `properties` -> `override` -> `readonly: yes` so that zrepl itself will set that property on the non-placeholder datasets it creates. For simplicitly `canmount` and `snapdir` can also be set there, although they could also be set at the `root_fs` level and inherited using `recv` -> `properties` -> `inherit`.
